lib/single-selection/useTreeNode.tsx

```diff
@@ -0,0 +1,168 @@
+import { useContext, FocusEvent, MouseEvent, KeyboardEvent } from "react";
+import isHotkey from "is-hotkey";
+
+import { TreeViewContextType, TreeViewContext } from "./tree-context";
+import { TreeActionTypes, TREE_ID } from "./tree-state";
+import {
+  getFirstNode,
+  getLastNode,
+  getNextByTypeahead,
+  getNextFocusable,
+  getPreviousFocusable,
+} from "./tree-traversal";
+import { TreeNodeMetadataType } from "lib/research/treeContext";
+
+export function useTreeNode(id: string): {
+  isOpen: boolean;
+  open: () => void;
+  close: () => void;
+  isFocused: boolean;
+  isSelected: boolean;
+  children: string[];
+  metadata: TreeNodeMetadataType;
+  getTreeNodeProps: () => {
+    ref: (current: HTMLElement | null) => void;
+    tabIndex: number;
+    ["aria-expanded"]: boolean;
+    ["aria-selected"]: boolean;
+    role: "treeitem";
+    onClick: (event: MouseEvent) => void;
+    onKeyDown: (event: KeyboardEvent) => void;
+    onBlur: (event: FocusEvent) => void;
+    onFocus: (event: FocusEvent) => void;
+  };
+  treeGroupProps: {
+    role: "group";
+  };
+} {
+  const { state, dispatch, elements } =
+    useContext<TreeViewContextType>(TreeViewContext);
+  const { focusableId } = state;
+
+  const isOpen = state.isOpen.get(id) ?? false;
+  const metadata = state.metadata.get(id) ?? {
+    name: "Untitled",
+    isFolder: false,
+  };
+  const children = state.children.get(id) ?? [];
+
+  return {
+    isOpen,
+    metadata,
+    children,
+    isFocused: state.focusedId === id,
+    isSelected: state.selectedId === id,
+    open: function () {
+      dispatch({ type: TreeActionTypes.OPEN, id });
+    },
+    close: function () {
+      dispatch({ type: TreeActionTypes.CLOSE, id });
+    },
+    getTreeNodeProps: () => ({
+      ref: function (current: HTMLElement | null) {
+        if (current) {
+          elements.current.set(id, current);
+        } else {
+          elements.current.delete(id);
+        }
+      },
+      ["aria-expanded"]: metadata.isFolder && isOpen,
+      ["aria-selected"]: state.selectedId === id,
+      role: "treeitem",
+      tabIndex: focusableId === id ? 0 : -1,
+      onClick: function (event: MouseEvent) {
+        // https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent/button
+        // is this the left mouse button / main mouse button
+        if (event.button === 0) {
+          event.stopPropagation();
+
+          if (metadata.isFolder) {
+            isOpen
+              ? dispatch({ type: TreeActionTypes.CLOSE, id })
+              : dispatch({ type: TreeActionTypes.OPEN, id });
+          }
+          dispatch({ type: TreeActionTypes.SET_FOCUSABLE, id });
+          dispatch({ type: TreeActionTypes.SELECT, id });
+        }
+      },
+      onKeyDown: function (event: KeyboardEvent) {
```

Comment:
just realized that I didn't called `props.onKeyDown` in here, which defeats the point of using a prop getter. I will fix that in this and other methods soon